Realm/RLMUtil.hpp

```diff
@@ -72,9 +74,12 @@ static inline T *RLMDynamicCast(__unsafe_unretained id obj) {
 
 template<typename T>
 static inline T *RLMNSNullToNil(T *obj) {
-    if (obj == NSNull.null) {
+    if (static_cast<id>(obj) == NSNull.null) {
         return nil;
     }
+    else if (RLMIsKindOfClass(object_getClass(obj), [RLMOptionalBase class])) {
+        return [static_cast<RLMOptionalBase *>(obj) underlyingValue];
```

Comment:
A rename so the method doesn't seem specific to handling `NSNull` would address my concern.